mdagent/tools/base_tools/analysis_tools/pca_tools.py

```diff
@@ -0,0 +1,370 @@
+        self.reduce_dim = None
+        self._sub_pcs = None
+        self.n_pcs = None
+
+    def _align_trajectory(self):
+        ###########MD TRAJ ALIGNMENT################
+        try:
+            self.traj.superpose(self.traj, frame=0, atom_indices=self.atom_indices)
+            return "Trajectory aligned to the first frame. "
+        except Exception as e:
+            print(f"Error aligning trajectory: {str(e)}")
+
+    def get_pc(self):
+        align_message = self._align_trajectory()
+        # self.pc = pca.PCA(
+        #    self.u,
+        #    select=self.selection,
+        #    align=True,
+        #    mean=None,
+        #    n_components=None,
+        #    verbose=True,
+        # ).run()
+        self.reduce_dim = self.pc.fit_transform(
+            self.traj.xyz[:, self.atom_indices].reshape(
+                self.traj.n_frames, len(self.atom_indices) * 3
+            )
+        )
+
+        return "PCA done" + align_message
+
+    def _sub_array_sum_to_m(self, arr, M):
+        result = []
+        sum = 0
+        for value in arr:
+            if sum >= M:
+                break
+            result.append(value)
+            sum += value
+        return result
+
+    def _get_number_pcs(self):
+        self._sub_pcs = self._sub_array_sum_to_m(
+            self.pc.explained_variance_ratio_, self.pc_percentage
+        )
+        self.n_pcs = len(self._sub_pcs)
+        if self.n_pcs > 3:
+            self.n_pcs = 3
+        return self.n_pcs
+
+    def _make_transformation(self):
+        self.transformed = self.reduce_dim[:, : self.n_pcs]
+
+    def _make_df(self):
+        self._get_number_pcs()
+        self._make_transformation()
+        self.pc_df = pd.DataFrame(
+            self.transformed, columns=["PC{}".format(i + 1) for i in range(self.n_pcs)]
+        )
+        self.pc_df["Time (ps)"] = self.pc_df.index * self.traj.timestep
+
+    def _cosine_content(self, pca_space, i):
+        """Measure the cosine content of the PCA projection.
+
+        The cosine content of pca projections can be used as an indicator if a
+        simulation is converged. Values close to 1 are an indicator that the
+        simulation isn't converged. For values below 0.7 no statement can be made.
+        If you use this function please cite [BerkHess1]_.
+        References
+        ----------
+        .. [BerkHess1] Berk Hess. Convergence of sampling in protein simulations.
+                    Phys. Rev. E 65, 031910 (2002).
+        """
+        t = np.arange(len(pca_space))
+        T = len(pca_space)
+        cos = np.cos(np.pi * t * (i + 1) / T)
+        return (
+            (2.0 / T)
+            * (scipy.integrate.simps(cos * pca_space[:, i])) ** 2
+            / scipy.integrate.simps(pca_space[:, i] ** 2)
+        )
+
+    def make_scree_plot(self):
+        extra_mess = ""
+        if not self.pc:
+            pc_mess = self.get_pc()
+            if "not aligned" in pc_mess:
```

Comment:
what is the purpose of this??

Reply:
Before making the PCA calculation, is important to align the trajectory (usually against the first frame), so it captures relative movement of atoms against themselves. But if is rotations, and or translations are not taking into account they can skew the analysis. Even then, the analysis can still be performed, but the results are less trustful, which is why I put that as part of the final message from the tool.

Reply:
Can we include it as a warning to the user, rather than a printed statement or message to the agent?

Reply:
This is a good idea

Reply:
alright, sorry for the misunderstanding haha I added a warning now, instead of just printing out.